Add data parameters

ElySeraidarian opened the pull request:
This PR aims to follow the process of solving issue #25.

Data parameters added:
- use_relative
- flipped
- order_descending
- order_tree
- add.significance
- add.expl.var

Current issues :
- confidence.level (RDAPlot) is a numeric between 0 and 1, so I'm using a sliderInput but even after adding step parameter there is only 0 and 1 that can be selected, moreover it doesn't seem to change anything on the plot either.

- shape_by (AbundanceDensityPlot) should be hidden when layout="density", the only method for this looks to be .conditionOnRadio but it should only work on radiobuttons & checkboxes (not selectInput). I have two solutions : creating a method for selectinputs or change the selectinput in radiobuttons?

Still to implement : 
- order_sample_by
- decreasing
- Visual params for both RDAPlot & RowTreePlot

Note : Have to be using the GitHub version of miaViz

RiboRings commented on R/class-AbundanceDensityPlot.R:
```diff
@@ -94,7 +101,14 @@ setMethod(".defineDataInterface", "AbundanceDensityPlot",
+        .checkboxInput.iSEE(x, field="flipped", label="Switch axis",
+            value=slot(x, "flipped")),
+        
+        .selectInput.iSEE(x, field="order_descending", label="Descending order",
```
How about verbalising the label: "Order decreasing"

RiboRings commented on R/class-AbundanceDensityPlot.R:
```diff
@@ -94,7 +101,14 @@ setMethod(".defineDataInterface", "AbundanceDensityPlot",
-            value=slot(x, "n"), min=1, max=nrow(se), step=1))
+            value=slot(x, "n"), min=1, max=nrow(se), step=1),
+        
+        .checkboxInput.iSEE(x, field="flipped", label="Switch axis",
```
"Switch axes"

RiboRings commented on R/class-AbundanceDensityPlot.R:
```diff
@@ -164,6 +178,8 @@ setMethod(".generateOutput", "AbundanceDensityPlot",
     args[["add_legend"]] <- deparse(slot(x, "add_legend"))
     args[["assay.type"]] <- deparse(slot(x, "assay.type"))
+    args[["flipped"]] <- deparse(slot(x , "flipped"))
+    args[["order_descending"]] <- deparse(slot(x, "order_descending"))
```
First, descending = FALSE and NA both give the same order. Could you check if this is the case only for Tengeler2020 or for also for other datasets? Or maybe it's explained in the function vignettes?

If FALSE and NA always give the same order, you can just reduce this parameter to a checkbox with TRUE and FALSE.

ElySeraidarian replied:
if NA, it chooses either TRUE or FALSE based on the value found in the TSE object (at least this is what documentation says)

RiboRings replied:
Ok then TRUE/FALSE checkbox is probably enough

RiboRings commented on R/class-AbundanceDensityPlot.R:
```diff
@@ -293,6 +327,11 @@ setMethod(".definePanelTour", "AbundanceDensityPlot", function(x) {
+            .conditionalOnRadio(
+                paste0(panel_name, "_layout"), "density",
+                iSEE:::.selectInputHidden(x, field="dots_shape_by",
+                    label="Shape by", choices=names(colData(se)),
```
Label: Shape dots by

RiboRings commented on R/class-AbundancePlot.R:
```diff
@@ -70,9 +72,18 @@ AbundancePlot <- function(...) {
+setMethod(".defineDataInterface", "AbundancePlot", function(x, se, select_info) {
+    panel_name <- .getEncodedName(x)
+          
+    list(.checkboxInput.iSEE(x, field="use_relative", label="Relative Values",
```
label: "Use relative values"

RiboRings commented on R/class-AbundancePlot.R:
```diff
@@ -208,6 +220,10 @@ setMethod(".definePanelTour", "AbundancePlot", function(x) {
+    .addSpecificTour(class(x)[1], "use_relative", function(panel_name) {
+        data.frame(rbind(c(element = paste0("#", panel_name,
+            "_use_relative"), intro = "Here, we can choose
+            whether or not to use relative values.")))})
```
"whether to use relative or absolute values"

RiboRings commented on R/class-RDAPlot.R:
```diff
@@ -89,14 +104,20 @@ RDAPlot <- function(...) {
     list(.selectInput.iSEE(x, field="dimred", label="Reduced dimension",
-        choices=reducedDimNames(se), selected=slot(x, "dimred")))
+            choices=reducedDimNames(se), selected=slot(x, "dimred")),
+        .sliderInput.iSEE(x, field="confidence.level", label="Confidence level",
```
I would rather use an inputNumeric like the number of taxa to show in the AbundanceDensityPlot panel. This is because users probably want to have more flexible choice within 0.95 to 0.999 range than for the rest

RiboRings commented on R/class-RDAPlot.R:
```diff
@@ -89,14 +104,20 @@ RDAPlot <- function(...) {
+            choices=reducedDimNames(se), selected=slot(x, "dimred")),
+        .sliderInput.iSEE(x, field="confidence.level", label="Confidence level",
+            min = 0, max = 1, step= 20, value=slot(x, "confidence.level")),
+        .checkboxInput.iSEE(x, field="add.significance", label="Variance in labels",
```
label: "Show variance in labels"

RiboRings commented on R/class-RDAPlot.R:
```diff
@@ -89,14 +104,20 @@ RDAPlot <- function(...) {
+            min = 0, max = 1, step= 20, value=slot(x, "confidence.level")),
+        .checkboxInput.iSEE(x, field="add.significance", label="Variance in labels",
+            value=slot(x, "add.significance")),
+        .checkboxInput.iSEE(x, field="add.expl.var", label="Variance in axes",
```
label: "Show variance in axes"

RiboRings commented on R/class-RDAPlot.R:
```diff
@@ -114,8 +135,8 @@ setMethod(".createObservers", "RDAPlot",
     .createProtectedParameterObservers(panel_name, c("dimred", "add.ellipse",
-        "colour_by", "vec.text", "add.vectors"), input=input, pObjects=pObjects,
-        rObjects=rObjects)
+        "colour_by", "vec.text", "add.vectors", "add.expl.var", "add.significance"),
```
Probably unresponsive to confidence.level, because you need to add it to the observers here

RiboRings commented on R/class-RDAPlot.R:
```diff
@@ -251,6 +276,22 @@ setMethod(".definePanelTour", "RDAPlot", function(x) {
             whether or not to add a box around labels.")))})
+    .addSpecificTour(class(x)[1], "confidence.level", function(panel_name) {
+        data.frame(rbind(c(element = paste0("#", panel_name,
+            "_confidence\\.level"), intro = "Here, we can choose
```
"Here, we can adjust the confidence level"

RiboRings commented on R/class-RDAPlot.R:
```diff
@@ -251,6 +276,22 @@ setMethod(".definePanelTour", "RDAPlot", function(x) {
+        data.frame(rbind(c(element = paste0("#", panel_name,
+            "_ellipse\\.alpha"), intro = "Here, we can choose
+            the ellipse opacity.")))})
+    .addSpecificTour(class(x)[1], "add.expl.var", function(panel_name) {
+        data.frame(rbind(c(element = paste0("#", panel_name,
+            "_add\\.expl\\.var"), intro = "Here, we can choose
+            whether or not to add variance in the labels.")))})
+    .addSpecificTour(class(x)[1], "add.significance", function(panel_name) {
+        data.frame(rbind(c(element = paste0("#", panel_name,
+            "_add\\.significance"), intro = "Here, we can choose
+            whether or not to add variance in the axes.")))})
     
```
"... whether or not to show the explained variance ..."

RiboRings commented on R/class-RDAPlot.R:
```diff
@@ -260,6 +301,8 @@ setMethod(".definePanelTour", "RDAPlot", function(x) {
         .selectInput.iSEE(x, field="add.ellipse", label="Ellipse style",
             choices = c("fill", "colour", "FALSE"),
             selected=slot(x, "add.ellipse")),
+        .sliderInput.iSEE(x, field="ellipse.alpha", label="Ellipse opacity",
+            min=0, max=1, step=10, value=slot(x, "ellipse.alpha")),
```
try reducing the step size for example to 0.1. I think this param is also missing in the .createObservers method

RiboRings commented on R/class-RowTreePlot.R:
```diff
@@ -87,6 +90,15 @@ RowTreePlot <- function(...) {
+setMethod(".defineDataInterface", "RowTreePlot", function(x, se, select_info) {
+  panel_name <- .getEncodedName(x)
+  
+  list(.checkboxInput.iSEE(x, field="order_tree", label="Tree order",
```
label: "Order tree"

RiboRings commented on R/class-RowTreePlot.R:
```diff
@@ -251,6 +264,11 @@ setMethod(".definePanelTour", "RowTreePlot", function(x) {
             whether or not to colour the nodes by a variable from the
             <code>rowData</code>. When active, the available options are listed
             and one of them can be selected.")))})
+    .addSpecificTour(class(x)[1], "order_tree", function(panel_name) {
+        data.frame(rbind(c(element = paste0("#", panel_name,
+            "_order_tree"), intro = "Here, we can choose
+            whether or use alphabetic order of taxonomic
+            levels to order the tree.")))})
```
"Here, we can order the tree alphabetically"

RiboRings commented on R/class-AbundanceDensityPlot.R:
```diff
@@ -293,6 +327,11 @@ setMethod(".definePanelTour", "AbundanceDensityPlot", function(x) {
                     label="Color dots by", choices=names(colData(se)), 
-                    selected=slot(x, "dots_colour_by"))))
+                    selected=slot(x, "dots_colour_by"))),
+            .conditionalOnRadio(
```
Maybe insteading of making shape_by depend on layout, you can just use the same approach as for "dots_colour". Later on we might change this into the visual option checkbox approach

RiboRings replied:
Could you add the parameter `add_legend` to the AbundanceDensityPlot panel? RowTreePlot can be used as a reference.

RiboRings reviewed (changes requested):
Great job! I added some points that can be improved. Also you might need to update the unit tests since there are new parameters now

ElySeraidarian commented:
I have corrected the issues and added the last data parameters.
I still have still have one issue for order_sample_by, we should also be able to use a class from the different taxonomy ranks , for example :
`plotAbundance(se, assay.type="relabundance", rank = "Phylum",
           order_rank_by="abund", order_sample_by = "Bacteroidetes")`
However when I try this I get an error as it is not in the abundance data.

Moreover, for the visual parameters for RDAPlot/RowTreePlot (tip_size_by/edge_size_by/node_size_by). It is needed to perform some modification to the data pefore plotting so I don't know if this is something to be added (also how).
For example : 
`data(GlobalPatterns)
altExps(GlobalPatterns) <- splitByRanks(GlobalPatterns)
altExp(GlobalPatterns,"Genus") <- addPerFeatureQC(altExp(GlobalPatterns,"Genus"))
rowData(altExp(GlobalPatterns,"Genus"))$log_mean <-
  log(rowData(altExp(GlobalPatterns,"Genus"))$mean)
rowData(altExp(GlobalPatterns,"Genus"))$detected <-
   rowData(altExp(GlobalPatterns,"Genus"))$detected / 100
top_genus <- getTopFeatures(altExp(GlobalPatterns,"Genus"),
                        method="mean",
                        top=100L,
                        assay.type="counts")
#
x <- altExp(GlobalPatterns,"Genus")
plotRowTree(x[rownames(x) %in% top_genus,],
            tip_colour_by = "log_mean",
            tip_size_by = "detected")`

RiboRings commented:
Nice progress!

For the size parameters of RDAPlot and RowTreePlot, we can just provide the parameter without adding code. Note that a simple example like this also works:

```
library(mia)
library(miaViz)
data("Tengeler2020", package = "mia")

tse <- addPerFeatureQC(tse)
plotRowTree(tse, tip_size_by = "mean")
```

If you think that some more advanced usage should be explained (like the code you wrote above), the right place is in `vignettes/` as a separate Rmd file. For example, see the Articles section in the [iSEE documentation](https://isee.github.io/iSEE/).

ElySeraidarian commented:
I have added example for Row data so I'll explain what's the issue:
 `choices=unique(rowData(se)$Phylum)`
For example here I've added this line as choices for the selectInput, to be able to use it and that the app doesn't crash you need to firstly select "Phylum" rank in the visual parameters and then it will work. So here is the issue : we should update this list each time the rank changes so that it follows the right possibl values

ElySeraidarian commented:
Now all the visual parameters should have been added in both RDAPlot & RowTreePlot (maybe there can be a better way to sort these parameters in RowTreePlot, I can rework this).

RiboRings commented:
> Now all the visual parameters should have been added in both RDAPlot & RowTreePlot (maybe there can be a better way to sort these parameters in RowTreePlot, I can rework this).

Looks good!

In the RDAPlot panel, the confidence levels and the visual parameters listed in the checkbox seem unresponsive to changes. Could you check what's going on?

ElySeraidarian commented:
> > Now all the visual parameters should have been added in both RDAPlot & RowTreePlot (maybe there can be a better way to sort these parameters in RowTreePlot, I can rework this).
> 
> Looks good!
> 
> In the RDAPlot panel, the confidence levels and the visual parameters listed in the checkbox seem unresponsive to changes. Could you check what's going on?

For me it is responsive to changes, the minor issues I can are : 
- panels for sliderInputs doesn't not indicate the right place
- Labels on plot slightly move when opening each checkbox (because the size window is changed so it also changes the plot dimensions)

ElySeraidarian commented:
By the way, I have spotted another issue for the order_sample_by : 
if you try to change the taxonomic rank and access to Column data, the app will crash (looks like this is  because there is a conflict between order_sample_by rank and the taxonomic rank selected, e.g. selecting in row data a family name from the taxonomic rank you'd like to change then changing to Column data will work fine)

ElySeraidarian commented:
> By the way, I have spotted another issue for the order_sample_by : if you try to change the taxonomic rank and access to Column data, the app will crash (looks like this is because there is a conflict between order_sample_by rank and the taxonomic rank selected, e.g. selecting in row data a family name from the taxonomic rank you'd like to change then changing to Column data will work fine)

Fixed.

RiboRings commented:
> Now all the visual parameters should have been added in both RDAPlot & RowTreePlot (maybe there can be a better way to sort these parameters in RowTreePlot, I can rework this).

Ideally, we would like to adjust, say, line colour, while keeping line size unchanged. But in the current configuration one cannot activate colour by without also activating shaping by, or size by. Probably the checkbox solution as in the RDAPlot panel could fix this.

ElySeraidarian commented:
I have made changes for checkboxes but there are still issues : 
- After you have enabled the checkbox, for some reasons even if you uncheck, the changes will be kept
- If enabling/disabling several times in a row checkboxes, app will crash stating that length of if statement is too long

RiboRings commented:
> I have made changes for checkboxes but there are still issues :
> 
> * After you have enabled the checkbox, for some reasons even if you uncheck, the changes will be kept
> * If enabling/disabling several times in a row checkboxes, app will crash stating that length of if statement is too long

At least the first issue is likely due to this: say you turn on colour, set it to blue, and turn it off. Once you turn it off, the conditional in the .generateOutput method will return FALSE so colour will not be changed (will stay blue). Thus, we might need to look how it is done originally in the source code of iSEE visual box, or come up with something smart ourselves.

ElySeraidarian commented:
> > I have made changes for checkboxes but there are still issues :
> > 
> > * After you have enabled the checkbox, for some reasons even if you uncheck, the changes will be kept
> > * If enabling/disabling several times in a row checkboxes, app will crash stating that length of if statement is too long
> 
> At least the first issue is likely due to this: say you turn on colour, set it to blue, and turn it off. Once you turn it off, the conditional in the .generateOutput method will return FALSE so colour will not be changed (will stay blue). Thus, we might need to look how it is done originally in the source code of iSEE visual box, or come up with something smart ourselves.

In the reducedDimPlot there is quite the same issue (except that sometimes there is a conditional radio button after turning on the checkbox so it can be disabled), but app shouldn't crash at least

RiboRings commented:
Right! Two things:

1. Any configuration is ok as long as aesthetics for different components (lines, tips and nodes) are independent from each other, that is, it is possible to select line colour/size/shape without affecting node colour/size/shape
2. The error occurs when two or more boxes are checked (for example, colour + size). So probably .conditionalOnCheckGroup can only take one input at a time. In the source code of visual box they seem to use something different, which is worth checking out: https://github.com/iSEE/iSEE/blob/8f32b92fb8343bc77a16988cc3afc70d65b961fb/R/interface_visual.R#L70

ElySeraidarian commented:
Everything should be working I believe now, let me know if you find anything wrong still

RiboRings commented:
Thank you! We are getting closer.

We would like also colour, size and shape of, say, node to be independent from each other, so that we can colour nodes without turning on their size. How about the following configuration? (sorry for switching again)

[ ] Color  [ ] Shape  [ ] Size

Color by:
[ ] line  [ ] tip  [ ] node

Shape by:
[ ] tip  [ ] node

Size by:
[ ] line  [ ] tip  [ ] node

By default, nothing is active. When you turn on for example colour, you will see options line, tip and node. You can choose to colour only one of them, or more, but once you turn off the corresponding checkbox, say line, the colour of line should go back to default (turn off line colour).

Do you think this can be a good solution? Let me know if you need any help.

Btw, currently colour is working only for edge and not for line and tip.

RiboRings commented:
Also if you want maybe you could add yourself to the contributors in DESCRIPTION?

ElySeraidarian commented:
The rework has been made, there shouldn't be any crash and should be operational.
The only thing is that when desactivating a checkbox it won't disable instantly the parameter (will do it when activating another checkbox). I don't if this can be fixed somehow

ElySeraidarian commented:
> The rework has been made, there shouldn't be any crash and should be operational. The only thing is that when desactivating a checkbox it won't disable instantly the parameter (will do it when activating another checkbox). I don't know if this can be fixed somehow

Like it does not refresh the plot when desactivating a checkbox which is probably the reason behind

RiboRings commented:
Ok, I think it's good to go. If we find a solution to turn off aesthetics we can add it in the future. Let's fix the checks and then we can merge.

ElySeraidarian commented:
> Ok, I think it's good to go. If we find a solution to turn off aesthetics we can add it in the future. Let's fix the checks and then we can merge.

Tests somehow fail in the github workflow but all tests were successful locally, probably takes the previous tests version

RiboRings commented:
Tests are probably failing because we are using the GitHub version of miaViz. @TuomasBorman, when will it be pushed to bioconductor devel?

TuomasBorman commented:
There are big changes both in mia and miaViz. I pushed now, but there might occur some issues since miaViz is not yet synced with mia
